### PR TITLE
update to pydantic v2

### LINKED ISSUE #50: Refactor CustomDateField
The current implementation is a bit hacky, and would not work with pydantic v2. It could be made into a `datetime` object, which Pydantic can parse from various formats,making it easier to create a message if not using a string. There should still however be a manual check when parsed from a string that it fits the ISO standard. Some custom logic (or someting in dateutil?) may be needed to turn the datetime into the proper format

### PR DESCRIPTION
**Pull request by ameyarao98**
Closes #50 and #46 

Note: Tests currently fail because the `AnyUrl` type always appends a `/` at the end:

```
>>> from pydantic import AnyUrl, BaseModel
>>> class x(BaseModel):
...     a:AnyUrl
... 
>>> x(a="https://tally.xyz")
x(a=Url('https://tally.xyz/'))
```

Either find a way to modify it not to do that or go for a simple string instead
https://github.com/pydantic/pydantic/issues/7186

### CODE REVIEW

**Comment by sbihel** [comment]
> Tests currently fail because the AnyUrl type always appends a / at the end

Related to that, it's the reason why there was a custom date type, to enable transitivity. Mainly because the verification method re-serialises the message before verifying the method. Which is arguably a wasteful process.

I see a few alternatives:
- keep wrapper types for fields that have multiple representations for the same thing;
- keep the original message in `SiweMessage`;
- with the introduction of a `from_message` method, it could do the verification before deserialising; or
- split the `SiweMessage` type into `SiweMessageBuilder` and `SignedSiweMessage`.

The last two are kind of the same, and I think it's probably the best way to go about it, JWT-style verification.

Do you have an opinion?

File: siwe/siwe.py [hunk 1]
```diff
@@ -7,11 +7,17 @@
 from typing import Any, Dict, Iterable, List, Optional, Union
 
 import eth_utils
-from dateutil.parser import isoparse
```
Comment (xcaptain) [inline siwe/siwe.py:10]: if `isoparse` is no longer used, I think you can also remove `python-dateutil` from `pyproject.toml`

File: tests/test_siwe.py [hunk 2]
```diff
@@ -1,12 +1,13 @@
 import json
+from datetime import datetime
 
 import pytest
 from dateutil.parser import isoparse
```
Comment (xcaptain) [inline tests/test_siwe.py:5]: I think you can use `datetime.fromisoformat` to replace `isoformat`

**Comment by xcaptain** [comment]
If the `AnyUrl` thing got fixed, please also upgrade `eth-account`, the current `0.8.0` is old

**Comment by ameyarao98** [comment]
> Do you have an opinion?

Just seeing this but I agree that `split the SiweMessage type into SiweMessageBuilder and SignedSiweMessage.` is the right way to go. I do think that that might be beyond the scope of the PR however,. For now I think making `AnyUrl` a string instead might be better, since I did have a similar issue with another project and I concluded that the benefits of Pydantic validating URL structure are minimal

File: siwe/siwe.py (hunk 1 above)
Comment (ameyarao98) [reply siwe/siwe.py:10]: deptry will complain about that in the CI if it an issue

**Comment by sbihel** [comment]
Closed by #59

**Pull request closed without merging**